### ISSUE TITLE
C_GetSlotList() can fail with pkcs11-tool -L and no card

### ISSUE THREAD

**Issue by vjardin** [p11-glue/p11-kit#257]
When pkcs11-tool -L (or pkcs11-dump slotinfo) is called, we get the error :
```
(p11-kit:173726) rpc_C_GetSlotList: ret: 336
Fatal: C_GetSlotList-CKR_BUFFER_TOO_SMALL
```
because of:
https://github.com/p11-glue/p11-kit/blob/6b33efaa82848bac7fd73888963a106cf4c03151/p11-kit/rpc-client.c#L871

In fact, count = 0 for the 1st call, then because of 0 slot, count = 0 for the 2nd call.
slot_list = NULL for the 1st call, then slot_slit = 0xA1pointer1A for the 2nd call.
However, for such 2nd call, since count = 0, rpc-server.c gets a slot_list = NULL over the RPC instead of some dummy pointers.
Then, the pkcs11 driver's C_GetSlotList() returns an error (CKR_BUFFER_TOO_SMALL for my case).
The intent of pkcs11-tool or pkcs11-dump for this 2nd call is to get a CKR_OK even if there is not slot however because of the RPC framework, rpc-server.c cannot guess this intent.

(fix to follow)

**Comment by vjardin** [comment]
A possible hack (or fix) could be the following:
```
diff --git a/p11-kit/rpc-client.c b/p11-kit/rpc-client.c
index e202e37..c6d9242 100644
--- a/p11-kit/rpc-client.c
+++ b/p11-kit/rpc-client.c
@@ -864,8 +864,20 @@ rpc_C_GetSlotList (CK_X_FUNCTION_LIST *self,
                    CK_SLOT_ID_PTR slot_list,
                    CK_ULONG_PTR count)
 {
+       CK_SLOT_ID _slot_list;
        return_val_if_fail (count, CKR_ARGUMENTS_BAD);
 
+#if 1
+       /*
+        * XXX: some perform a call with a list, still requesting 0 element
+        * doing so will lead to a CKR_BUFFER_TOO_SMALL !!! with some pkcs11
+        * drivers.
+        */
+       if (slot_list && count && (*count == 0)) {
+               slot_list = &_slot_list; /* we cannot use slot_list */
+               *count = 1;
+       }
+#endif
        BEGIN_CALL_OR (C_GetSlotList, self, (*count = 0, CKR_OK));
                IN_BYTE (token_present);
                IN_ULONG_BUFFER (slot_list, count);

```

**Comment by vjardin** [comment]
I feel that there is an issue with the semantic of the RPC for any cases when count = 0 but the pointer is something. TBD: what should we do ?

**Comment by vjardin** [comment]
With this fix, I can get it working then:

> $ pkcs11-dump slotlist p11-kit-client.so 
> pkcs11-dump 0.3.4 - PKI Cryptoki token dump
> Written by Alon Bar-Lev
> 
> Copyright (C) 2005-2006 Alon Bar-Lev.
> This is free software; see the source for copying conditions.
> There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
> 
> No slot found

**Comment by vjardin** [comment]
(analysis)
The semantical issue comes from:
```
proto_read_ulong_buffer (p11_rpc_message *msg,
                         CK_ULONG_PTR *buffer,
                         CK_ULONG *n_buffer)
[...]
        if (!p11_rpc_buffer_get_uint32 (msg->input, &msg->parsed, &length))
                return PARSE_ERROR;
        *n_buffer = length;
        *buffer = NULL;
        /* If set to zero, then they just want the length */
        if (length == 0)
                return CKR_OK;
        *buffer = p11_rpc_message_alloc_extra (msg, length * sizeof (CK_ULONG));
[...]
```

so, it is impossible to allocate (`p11_rpc_message_alloc_extra()`) a `*buffer` of at least 1 element when length is 0. So, the semantic of the **oasis** standard is changed thru the RPC framework:
1. a call with f(xyz, 0x1ApointerA1, 0) becomes,
2. f(xyz, NULL, 0);

1-> means return the elements into 0x1ApointerA1, if 0 is not enough, Oasis states that one should return CKR_BUFFER_TOO_SMALL
2-> means return the number of element so for a subsequent call once could allocate the proper number of entries.

**Comment by vjardin** [comment]
On the other hand (client side that sends the RPC request), we have:

```
p11_rpc_message_write_ulong_buffer (p11_rpc_message *msg,
                                    CK_ULONG count)
[...]
        /* Make sure this is in the right order */
        assert (!msg->signature || p11_rpc_message_verify_part (msg, "fu"));
        p11_rpc_buffer_add_uint32 (msg->output, count);
```

which is called by` IN_ULONG_BUFFER (slot_list, count)` that sets only 'count' but not slot_list.

so it had become impossible for the peer side to guess that there was something for the pointers. I believe that, for all fhe "f_" cases, we need  to add a pointer length too.

Maybe, we could add the following arguments on both side of the RPC:
```
#define IN_ULONG_BUFFER(arr, len) \
        if (len == NULL) \
                { _ret = CKR_ARGUMENTS_BAD; goto _cleanup; } \
-       if (!p11_rpc_message_write_ulong_buffer (&_msg, arr ? *len : 0)) \
+       if (!p11_rpc_message_write_ulong_buffer (&_msg, arr ? sizeof(*arr): 0, *len)) \
                { _ret = CKR_HOST_MEMORY; goto _cleanup; }
```

The same should be done for all the *_buffer cases, it means:
- IN_BYTE_BUFFER,
- IN_ULONG_BUFFER,
- IN_ATTRIBUTE_BUFFER.

**OR (TBC)**, should we use `p11_rpc_message_write_ulong_array()` instead of `p11_rpc_message_write_ulong_buffer()` ?

**Comment by vjardin** [comment]
In fact, we should not use the `*_array()` methods because the content for the vector that will return each slot is 'memory'-allocated, but its content is meaningless. We should not try to send dummy memory content over the RPC.

**Comment by vjardin** [comment]
For the records, enclosed a WIP patch for some comments. Later on, I'll provide it thru a proper pull request.

[ulong_buffer_patch.txt](https://github.com/p11-glue/p11-kit/files/3749300/ulong_buffer_patch.txt)

**Comment by vjardin** [comment]
this bug is solved since the merge of the pull request #259 
Let's close it.
